### PR TITLE
Added a catch-all re_path to deal with non-existent resources which a…

### PR DESCRIPTION
**Pull request by briddo**
…re requested

### CODE REVIEW

File: nautobot/core/tests/test_api.py [hunk 1]
```diff
@@ -15,3 +17,9 @@ def test_status(self):
         response = self.client.get("{}?format=api".format(url), **self.header)
 
         self.assertEqual(response.status_code, 200)
+
+    def test_non_existent_resource(self):
+        url = reverse("api-root")
+        response = self.client.get(f"{url}/non-existent-resource-url/", **self.header)
```
Comment (glennmatthews) [inline nautobot/core/tests/test_api.py:23]: I'd expect to also see a `self.assertEqual(response.status_code, 404)` here, I think?

File: nautobot/core/api/views.py [hunk 2]
```diff
@@ -552,3 +553,7 @@ def execute_graphql_request(self, request, data, query, variables, operation_nam
             return document.execute(**options)
         except Exception as e:
             return ExecutionResult(errors=[e], invalid=True)
+
+
+def json404(request):
+    return JsonResponse({"status_code": 404, "error": "The resource was not found"})
```
Comment (glennmatthews) [inline nautobot/core/api/views.py:559]: Perhaps?

```suggestion
    return JsonResponse({"error": "The resource was not found"}, status=404)
```

File: nautobot/core/api/views.py [hunk 3]
```diff
@@ -552,3 +553,7 @@ def execute_graphql_request(self, request, data, query, variables, operation_nam
             return document.execute(**options)
         except Exception as e:
             return ExecutionResult(errors=[e], invalid=True)
+
+
+def json404(request):
```
Comment (jathanism) [inline nautobot/core/api/views.py:558]: Sorry to come in at the 11th hour here, but I'd like to see this expanded to be more of `HttpOrJson404` pattern.

I think a better solution would be:

- Define [`handler404`](https://docs.djangoproject.com/en/4.0/topics/http/views/#customizing-error-views) inside of `nautobot.core.urls` to point to the custom view defined here.
- Inside of the custom view, inspect `request.path`:
  - If it starts with `/api/` raise a [`rest_framwork.exceptions.NotFound` from DRF](https://www.django-rest-framework.org/api-guide/exceptions/#notfound) which returns a JSON 404 response.
   - Else return the [default `django.views.defaults.page_not_found` 404 view from Django](https://docs.djangoproject.com/en/4.0/ref/views/#the-404-page-not-found-view) which returns a HTTP 404 response.

**Review by jathanism — CHANGES_REQUESTED** [review]

File: nautobot/core/api/views.py (hunk 3 above)
Comment (jathanism) [reply nautobot/core/api/views.py:558]: The reason for this is it leaves the HTTP 404 responses the same as they are now, and makes sure that any JSON 404 responses are identical to ones that aren't customized.

Comment (briddo) [reply nautobot/core/api/views.py:558]: Good idea! Will do now...

Comment (briddo) [reply nautobot/core/api/views.py:558]: I modified the code as suggested, but found that raising NotFound actually returns a 500 response (weirdly) so stuck with a JsonResponse.

**Review by glennmatthews — APPROVED** [review]

**Review by jathanism — APPROVED** [review]

**Pull request merged**